First steps towards generalization in reimbursiments description

silviodc opened the pull request:
It is the inclusion of https://github.com/datasciencebr/serenata-de-amor/pull/238 in Rosie.
Since the built classifier has 91% of accuracy, maybe it can be included in the set of classifiers for the chamber of deputies.

PS:
1) I uploaded the ML model (40 MB) together to code. I guess it is not the best practice, maybe it must be downloaded from the serenata-toolbox, however i don't have access to amazon to upload the model  :/ ( * Probably we have to change it in order to have a better architecture)

2) The code works fine in the Test:
![screenshot from 2017-07-20 14-01-53](https://user-images.githubusercontent.com/7073485/28416678-4d1b34d0-6d55-11e7-9663-bd2180a418ed.png)
However i couldn't run it in the real reimbursements. Rosie is consuming all my RAM (5 Gb) while the reimbursements are downloaded, maybe someone could test it to me :+1: 

3) Sorry for my Python code. I just started to build something in Python there 4 months. I will appreciate some tips to improve the code.

4) As you can see in the print the TensorFlow i included is the basic one. I kept it since i don't know whether the machine where Rosie is running allows other configurations. (*We also can change it)

Commit:
First steps towards generalization in reimbursiments description

luzfcb commented:
A comment about 40 MB of binary file (ML model)
Commit a binary file on git repository is a bad idea. I have problems with this decision, because in https://github.com/pythonclub/pythonclub.github.io we had commited too many images files, about 3MB in total. The result was that, over time, the size of the repository increases exponentially, even though I did not include any other new binary files. (So far, the pythonclub repository is close to 216.33MB) Removing these files forces me to rewrite a whole git history, and this would break all branchs and forks. So I can not do that.

In the last year, github has added support for Git LFS, which was apparently created to solve problems like this.

https://git-lfs.github.com/
https://help.github.com/articles/versioning-large-files/

Commit:
Merge branch 'master' into master

Commit:
Update .travis.yml

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

trying to save the pdf and png in a temporary folder

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update .travis.yml

Commit:
Update .travis.yml

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update reimbursement_generalization.py

Commit:
Update .travis.yml

Commit:
Update .travis.yml

Commit:
Update .travis.yml

Commit:
Update .travis.yml

jtemporal commented on Dockerfile:
```diff
@@ -1,22 +1,22 @@
 COPY rosie.py ./
 COPY rosie ./rosie
 COPY config.ini.example ./
+COPY config.ini ./
```
This doesn't need to be here since Rosie's `setup` file creates `config.ini` from the example file copied in the previous line 😉 

You can go ahead and delete this line.

silviodc replied:
Done, i removed it. Strange the first time i tried to build the image i got error without it... Any way now its ok

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+from keras.layers import Activation, Dropout, Flatten, Dense
+from keras import backend as K
+from keras.callbacks import ModelCheckpoint
+from keras.models import load_model
```
You can follow [PEP-8 imports section](https://www.python.org/dev/peps/pep-0008/#imports) to guide you when organizing your imports. For example, when doing `from ... import ...` you can separate the second block with comma like: `from keras.models import Sequential, load_model`.

Note that you should also group your imports:
```Imports should be grouped in the following order:

1. standard library imports
1. related third party imports
1. local application/library specific imports

You should put a blank line between each group of imports.
```

Also on that note, you can make use of tools to help you automatically organize your imports like [isort](https://pypi.python.org/pypi/isort/).

silviodc replied:
Done! Now they look beautiful :D

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+        The year the expense was generated.
+    """
+
+    COLS = ['applicant_id',
```
Don't be afraid to use descriptive names like `COLUMNS` for your variables. It will make your code clearer to read down the road.

silviodc replied:
In fact i tried to follow the code of other classifiers. Ctrl+C > Ctrl+V
I changed it to COLUMNS ;)

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+        nb_train_samples = sum([len(files) for r, d, files in os.walk(train_data_dir)])
+        nb_validation_samples = sum([len(files) for r, d, files in os.walk(validation_data_dir)])
+
+        print('no. of trained samples = ', nb_train_samples, ' no. of validation samples= ',nb_validation_samples)
```
Long lines [aren't a good thing](https://www.python.org/dev/peps/pep-0008/#maximum-line-length). This print could be like the following in other to avoid the extra long line:
```
print('no. of trained samples = ', nb_train_samples,
      ' no. of validation samples= ', nb_validation_samples)
```
Extra space was missing there.

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+
+    img_width, img_height = 300, 300
+
+    def train(self,train_data_dir,validation_data_dir,save_dir):
```
Pay extra attention to spaces after commas, they help make your code easier on the eyes 😉

This method would be nicer like this:
```
def train(self, train_data_dir, validation_data_dir, save_dir):
```

silviodc replied:
I changed the code using this tool: http://pep8online.com/checkresult
PEP8 online
Check your code for PEP8 requirements

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+
+        model.add(Conv2D(64, (3, 3)))
+        model.add(Activation('relu'))
+        model.add(MaxPooling2D(pool_size=(2, 2)))
```
I noted that you make the same process here 3 times with very little difference between parameters used. Maybe you can explain a little why is that necessary?

silviodc replied:
I included a brief description and a link to explain it: http://deeplearning.net/tutorial/lenet.html

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+                      optimizer='rmsprop',
+                      metrics=['accuracy'])
+
+        #this is the augmentation configuration we will use for training
```
[PEP-8 inline comments](https://www.python.org/dev/peps/pep-0008/#inline-comments) state that:
>They (inline comments) should start with a # and a single space.

`# This is the augmentation configuration we will use for training` looks way better don't you think?

silviodc replied:
Done, using the tool

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+            rescale=1. / 255,
+            shear_range=0.2,
+            zoom_range=0.2,
+            horizontal_flip=False)#As you can see i put it as FALSE and on link example it is TRUE
```
[PEP-8 inline comments](https://www.python.org/dev/peps/pep-0008/#inline-comments) state that:
>An inline comment is a comment on the same line as a statement. Inline comments should be separated by at least two spaces from the statement. They should start with a # and a single space.

And be careful on the line length here too.

jtemporal replied:
Also, I didn't quite get what you meant here.

silviodc replied:
It was a copy and past from serenata. Now i included the reason

I put horizontal_flip as FALSE because we can not handwrite from right to left in Portuguese

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+            horizontal_flip=False)#As you can see i put it as FALSE and on link example it is TRUE
+        #Explanation, there no possibility to write in a reverse way :P
+
+        #this is the augmentation configuration we will use for testing:
```
Same thing as inline comments, these should start with `#` followed by a single space.

jtemporal replied:
This is applicable to all other comments you made in this file ;)

silviodc replied:
PEP8 requirements done, using it: http://pep8online.com/checkresult

jtemporal commented on rosie/chamber_of_deputies/classifiers/reimbursement_generalization.py:
```diff
@@ -0,0 +1,210 @@
+            class_mode='binary')
+
+        #It allow us to save only the best model between the iterations
+        checkpointer = ModelCheckpoint(filepath=save_dir+"weights.hdf5", verbose=1, save_best_only=True)
```
File paths should be built using `os.path.join` to avoid breaking the scripts when running Rosie in different systems

luzfcb replied:
@jtemporal Another good option with good API is [pathlib](https://docs.python.org/3/library/pathlib.html)

silviodc replied:
Done, i included the os.path.join

jtemporal reviewed (commented):
First comments on 💅 only. PEP-8 for the win ;)

Commit:
1) Review code for PEP8 requirements (using: http://pep8online.com/checkresult)
2) Include comments in the source code

Commit:
Merge commit '9396ecd1ab171c524827c2fd72349ac671f313ec'

Commit:
Include tests for training

Commit:
Include test to CORE

Commit:
Changing location of supervised models to settings

Commit:
Fixing integration between core and senate

Commit:
Fixing load supervised model

PEP8 over the changes

silviodc commented:
Main points after the last pull request:

1.  I deleted the Machine Learn model  of 40MB 👍  Now i configured the code to load the model  from some informed path or download from external source (the one i'm using).
2. I included tests to predict downloaded reimbursements using the proper ML model 
3. I included a test to create and to train a fake model.
4. I included a test to this classifier in TestCore class
5. I fixed the location of supervised machine learn models in `rosie.chamber_of_deputies.settings ` and `rosie.federal_senate.settings`. Therefore, new supervised models will be able to specific the files they have to load.
